tokio/src/net/tcp/split_owned.rs

```diff
@@ -0,0 +1,224 @@
+//! `TcpStream` owned split support.
+//!
+//! A `TcpStream` can be split into an `OwnedReadHalf` and a
+//! `OwnedWriteHalf` with the `TcpStream::split_owned` method.
+//! `OwnedReadHalf` implements `AsyncRead` while `OwnedWriteHalf`
+//! implements `AsyncWrite`.
+//!
+//! Compared to the generic split of `AsyncRead + AsyncWrite`, this specialized
+//! split has no associated overhead and enforces all invariants at the type
+//! level.
+
+use crate::future::poll_fn;
+use crate::io::{AsyncRead, AsyncWrite};
+use crate::net::TcpStream;
+
+use bytes::Buf;
+use std::error::Error;
+use std::mem::MaybeUninit;
+use std::net::Shutdown;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::{fmt, io};
+
+/// Owned read half of a [`TcpStream`], created by [`split_owned`].
+///
+/// [`TcpStream`]: TcpStream
+/// [`split_owned`]: TcpStream::split_owned()
+#[derive(Debug)]
+pub struct OwnedReadHalf(Arc<TcpStream>);
+
+/// Owned write half of a [`TcpStream`], created by [`split_owned`].
+///
+/// Note that in the `AsyncWrite` implemenation of this type, `poll_shutdown` will
+/// shut down the TCP stream in the write direction.
+///
+/// [`TcpStream`]: TcpStream
+/// [`split_owned`]: TcpStream::split_owned()
+#[derive(Debug)]
+pub struct OwnedWriteHalf(Arc<TcpStream>);
+
+pub(crate) fn split_owned(stream: TcpStream) -> (OwnedReadHalf, OwnedWriteHalf) {
+    let arc = Arc::new(stream);
+    let arc2 = Arc::clone(&arc);
+    (OwnedReadHalf(arc), OwnedWriteHalf(arc2))
+}
+
+pub(crate) fn reunite(
+    read: OwnedReadHalf,
+    write: OwnedWriteHalf,
+) -> Result<TcpStream, ReuniteError> {
+    if Arc::ptr_eq(&read.0, &write.0) {
+        drop(write);
```

Comment:
I think a common expectation with those split sockets is that dropping the writing half is equivalent to a close or shutdown of the socket. It seems like we don't this here - but maybe we should?

If we would do this, then reunite would not be allowed to call the destructor.

Reply:
That would make `reunite` more annoying to implement.

Reply:
Not only that. I think the reading half also couldn't be unmodified. If the `fd` gets closed it is not allowed for the reader to read/poll anymore. Instead it needs to read the close from a buffered state and may not access the `fd` anymore.

That's certainly a lot more complexity. But in all the scenarios where I used a separate reader and writer so far in other environments (Go, Java), it was necessary to have the functionality in order to prevent readers from being stuck and to implement a clean shutdown.

Now this might be different in Rust, because you can cancel an `async read` at any point in time - but it requires the read to be somewhere wrapped ina  `select` block which also waits on the closing event.

TBH: Not really sure what the best API is here.

Reply:
I admit that I'm not too familiar with similar apis in other languages, but I can see how it makes sense for dropping the writer to close the socket. It seems like it could be implemented like this:

1. When writer is dropped, call [`TcpStream::shutdown`][1], which I assume will abort any pending reads.
2. Make the write handle `#[repr(transparent)]` and transmute it into an `Arc` to skip the destructor on reunite.
3. Don't have any special destructor for the read half.

I don't really see any safe way of doing it.

[1]: https://docs.rs/tokio/0.2/tokio/net/struct.TcpStream.html#method.shutdown

Reply:
It should be possible to make it self by implementing drop along:

```rust
if self.shutdown_on_drop {
    self.0.shutdown();
}
```

and setting `shutdown_on_drop` in the writer to `false` before reunite. Downside is it increases the struct size.

Reply:
> 
> 
> It should be possible to make it safe by implementing drop along:
> 
> ```rust
> if self.shutdown_on_drop {
>     self.0.shutdown();
> }
> ```
> 
> and setting `shutdown_on_drop` in the writer to `false` before reunite. Downside is it increases the struct size.

Reply:
Ah, that's true. Which one do you prefer?

Reply:
I'm generally not too afraid of `unsafe` things, but in this case I would opt for the safe version. Don't think those 4-8 bytes hurt, given the general resource usage of a TCP stream.